dawn/src/dawn/SIR/SIR.h

```diff
@@ -302,89 +302,59 @@ struct Value : NonCopyable {
   /// Convert the value to string
-  std::string toString() const;
+  std::string toString() const { return std::string{typeToString(type_)}; }
 
   /// @brief Check if value is set
-  bool has_value() const { return valueImpl_->has_value(); }
+  bool has_value() const { return value_.has_value(); }
 
   /// @brief Get/Set the underlying type
-  TypeKind getType() const { return valueImpl_->getType(); }
+  TypeKind getType() const { return type_; }
 
   /// @brief Get the value as type `T`
   /// @returns Copy of the value
   template <class T>
   T getValue() const {
     DAWN_ASSERT(has_value());
-    DAWN_ASSERT_MSG(TypeInfo<T>::Type == valueImpl_->getType(), "type mismatch");
-    return *(T*)valueImpl_->get();
+    DAWN_ASSERT_MSG(getType() == TypeInfo<T>::Type, "type mismatch");
+    return std::get<T>(*value_);
   }
 
-  template <class T>
-  struct TypeInfo {
-    static const TypeKind Type;
-  };
-
   bool operator==(const Value& rhs) const;
   CompareResult comparison(const sir::Value& rhs) const;
 
   json::json jsonDump() const {
     json::json valueJson;
-    valueJson["type"] = Value::typeToString(valueImpl_->getType());
+    valueJson["type"] = Value::typeToString(getType());
     valueJson["isConstexpr"] = isConstexpr();
     valueJson["value"] = toString();
     return valueJson;
   }
 
-private:
-  struct ValueImplBase {
-    virtual ~ValueImplBase() {}
-    virtual void* get() = 0;
-    virtual TypeKind getType() const = 0;
-    virtual bool has_value() const = 0;
-  };
-
-  template <class T>
-  struct ValueImpl : public ValueImplBase {
-    T* ValuePtr;
-
-    ValueImpl() : ValuePtr(nullptr) {}
-    ValueImpl(const T& value) : ValuePtr(new T) { *ValuePtr = value; }
-    ~ValueImpl() {
-      if(ValuePtr)
-        delete ValuePtr;
-    }
-    void* get() override { return (void*)ValuePtr; }
-    TypeKind getType() const override { return TypeInfo<T>::Type; }
-    bool has_value() const override { return ValuePtr != nullptr; }
-  };
+  template <typename T> struct TypeInfo {};
 
-  bool isConstexpr_;
-  std::unique_ptr<ValueImplBase> valueImpl_;
+private:
+  std::optional<std::variant<bool, int, double, std::string>> value_;
+  bool is_constexpr_;
+  TypeKind type_;
 };
 
-template <>
-struct Value::TypeInfo<bool> {
-  static const Value::TypeKind Type = Value::Boolean;
+template <> struct Value::TypeInfo<bool> {
+ static const TypeKind Type = Boolean;
 };
 
-template <>
-struct Value::TypeInfo<int> {
-  static const Value::TypeKind Type = Value::Integer;
+template <> struct Value::TypeInfo<int> {
+  static const TypeKind Type = Integer;
 };
 
-template <>
-struct Value::TypeInfo<double> {
-  static const Value::TypeKind Type = Value::Double;
+template <> struct Value::TypeInfo<double> {
+  static const TypeKind Type = Double;
 };
 
-template <>
-struct Value::TypeInfo<std::string> {
-  static const Value::TypeKind Type = Value::String;
+template <> struct Value::TypeInfo<std::string> {
+  static const TypeKind Type = String;
```

Comment:
```suggestion
  static constexpr TypeKind Type = String;
```

Reply:
I had this originally, but in the context it's used the constexpr isn't helpful. I suppose it's still good form to have that.